crates/host/src/module.rs

```diff
@@ -1,38 +1,136 @@
+use crate::prelude::Instance;
 use holochain_wasmer_common::WasmError;
 use once_cell::sync::Lazy;
+use parking_lot::Mutex;
 use parking_lot::RwLock;
+use plru::MicroCache;
 use std::collections::HashMap;
+// use std::sync::atomic::AtomicU64;
+use bimap::BiMap;
 use std::sync::atomic::AtomicUsize;
 use std::sync::Arc;
 use wasmer::Cranelift;
 use wasmer::Module;
 use wasmer::Store;
 use wasmer::Universal;
 
+pub type CacheKey = [u8; 32];
+pub type PlruKeyMap = BiMap<usize, CacheKey>;
+pub type SerializedModule = Vec<u8>;
+
+trait PlruCache {
+    type Item;
+
+    fn plru_mut(&mut self) -> &mut MicroCache;
+    fn key_map(&self) -> &PlruKeyMap;
+    fn key_map_mut(&mut self) -> &mut PlruKeyMap;
+    fn cache(&self) -> &HashMap<CacheKey, Arc<Self::Item>>;
+    fn cache_mut(&mut self) -> &mut HashMap<CacheKey, Arc<Self::Item>>;
+
+    fn put_item(&mut self, key: CacheKey, item: Self::Item) -> Arc<Self::Item> {
+        let plru_key = self.plru_mut().replace();
+        // if there is something in the cache for this plru slot already drop it.
+        if let Some(stale_key) = self.key_map().get_by_left(&plru_key).cloned() {
```

Comment:
I don't think you need to clone then take a reference?

Reply:
@freesig doing this to get around:

```
error[E0502]: cannot borrow `*self` as mutable because it is also borrowed as immutable
  --> /home/thedavidmeister/holochain-wasmer/crates/host/src/module.rs:33:13
   |
32 |         if let Some(stale_key) = self.key_map().get_by_left(&plru_key) {
   |                                  ---- immutable borrow occurs here
33 |             self.cache_mut().remove(stale_key);
   |             ^^^^^^^^^^^^^^^^        --------- immutable borrow later used here
   |             |
   |             mutable borrow occurs here
```

it is only cloning a `[u8; 32]` so i figured the overhead would be negligible